File: src/Java/myservicebus/src/main/java/com/myservicebus/ConsumerDefinition.java
package com.myservicebus;

/**
 * Consumer-specific policy applied before a registration is materialized into
 * bus topology.
 *
 * @param <TConsumer> consumer implementation type
 */
public class ConsumerDefinition<TConsumer> {
    private final EndpointDefinition endpoint;

    public ConsumerDefinition() {
        this(new EndpointDefinition());
    }

    public ConsumerDefinition(EndpointDefinition endpoint) {
        if (endpoint == null) {
            throw new IllegalArgumentException("endpoint must not be null");
        }
        this.endpoint = endpoint;
    }

    public EndpointDefinition getEndpoint() {
        return endpoint;
    }

    public String getEndpointName() {
        return endpoint.getName();
    }

    public ConsumerDefinition<TConsumer> endpointName(String value) {
        endpoint.name(value);
        return this;
    }

    public Integer getConcurrentMessageLimit() {
        return endpoint.getConcurrentMessageLimit();
    }

    public ConsumerDefinition<TConsumer> concurrentMessageLimit(int value) {
        endpoint.concurrentMessageLimit(value);
        return this;
    }

    public Integer getPrefetchCount() {
        return endpoint.getPrefetchCount();
    }

    public ConsumerDefinition<TConsumer> prefetchCount(int value) {
        endpoint.prefetchCount(value);
        return this;
    }
}

File: src/MyServiceBus/ConsumerDefinition.cs
namespace MyServiceBus;

/// <summary>
/// Describes consumer-specific policy before it is materialized into bus topology.
/// </summary>
public interface IConsumerDefinition
{
    Type ConsumerType { get; }

    EndpointDefinition Endpoint { get; }

    string? EndpointName { get; }

    int? ConcurrentMessageLimit { get; }
}

/// <summary>
/// Defines the endpoint identity and execution policy associated with a consumer.
/// </summary>
/// <typeparam name="TConsumer">The consumer implementation type.</typeparam>
public class ConsumerDefinition<TConsumer> : IConsumerDefinition, IConsumerConfigurator<TConsumer>
    where TConsumer : class, IConsumer
{
    public ConsumerDefinition()
        : this(new EndpointDefinition())
    {
    }

    public ConsumerDefinition(EndpointDefinition endpoint)
    {
        ArgumentNullException.ThrowIfNull(endpoint);
        Endpoint = endpoint;
    }

    public Type ConsumerType => typeof(TConsumer);

    public EndpointDefinition Endpoint { get; }

    public string? EndpointName
    {
        get => Endpoint.Name;
        set => Endpoint.Name = value;
    }

    public int? ConcurrentMessageLimit
    {
        get => Endpoint.ConcurrentMessageLimit;
        set => Endpoint.ConcurrentMessageLimit = value;
    }

    public ushort? PrefetchCount
    {
        get => Endpoint.PrefetchCount;
        set => Endpoint.PrefetchCount = value;
    }
}

internal sealed class ConsumerRegistrationConfigurator<TConsumer>(Topology.ConsumerDefinitionModel definition)
    : IConsumerRegistrationConfigurator<TConsumer>
    where TConsumer : class, IConsumer
{
    public Topology.ConsumerDefinitionModel Definition { get; } = definition;
}
